psi4/src/psi4/scfgrad/response.cc

```diff
@@ -4399,52 +4407,64 @@ void USCFDeriv::JK_deriv2(std::shared_ptr<JK> jk, int mem,
         }
         for (int a = 0; a < nA; a++) {
-            psio_address next_Sij= psio_get_address(PSIO_ZERO,(A + a) * (size_t) n2occ * n2occ * sizeof(double));
-            psio_->read(PSIF_HESS,Sij_2,(char*)Sij2p[0], static_cast<size_t> (n2occ)*n2occ*sizeof(double),next_Sij, &next_Sij);
-            C_DGEMM('N','N',nso,n2occ,n2occ,1.0,C2op[0],n2occ,Sij2p[0],n2occ,0.0,R[2*a+1]->pointer()[0],n2occ);
+            psio_address next_Sij= psio_get_address(PSIO_ZERO,(A + a) * (size_t) nbocc * nbocc * sizeof(double));
+            psio_->read(PSIF_HESS,"Sij^A_b",(char*)Sijbp[0], static_cast<size_t> (nbocc)*nbocc*sizeof(double),next_Sij, &next_Sij);
+            C_DGEMM('N','N',nso,nbocc,nbocc,1.0,Cbop[0],nbocc,Sijbp[0],nbocc,0.0,R[2*a+1]->pointer()[0],nbocc);
+            Dx[2*a+1] = linalg::doublet(L[2*a+1], R[2*a+1], false, true);
+            Dx[2*a+1]->hermitivitize();
         }
 
         jk->compute();
         if(functional_->needs_xc()) {
             potential_->compute_Vx(Dx, Vx);
         }
 
-
         for (int a = 0; a < nA; a++) {
+            // Now construct the matrices. Both the alpha and beta term sum the two J terms.
+            // When we're done, J[2*a] will be the running alpha result,and J[2*a+1] the beta.
+            // For the convenience of other Hessian tech, all terms have an "extra" -1.
+            J[2*a]->add(J[2*a+1]);
+            J[2*a]->scale(-1);
+            J[2*a+1]->copy(J[2*a]);
+
+            // Subtract exchange, if any.
+            if (Kscale) {
+                K[2*a]->scale(Kscale);
+                K[2*a+1]->scale(Kscale);
+                J[2*a]->add(K[2*a]);
+                J[2*a+1]->add(K[2*a+1]);
+            }
 
-            // Add the alpha J contribution to G
-            C_DGEMM('N','N',nso,n1occ,nso,1.0,J[2*a]->pointer()[0],nso,C1op[0],n1occ,0.0,Tp[0],n1occ);
-            C_DGEMM('T','N',nmo,n1occ,nso,-1.0,C1p[0],nmo,Tp[0],n1occ,0.0,Up[0],n1occ);
-            T->zero();
+            // Add DFT, if any.
+            if (functional_->needs_xc()) {
+                // Symmetrize the result, just to be safe.
+                Vx[2*a]->hermitivitize();
+                Vx[2*a+1]->hermitivitize();
+                J[2*a]->subtract(Vx[2*a]);
+                J[2*a+1]->subtract(Vx[2*a+1]);
+            }
 
-            C_DGEMM('N','N',nso,n1occ,nso,1.0,J[2*a+1]->pointer()[0],nso,C1op[0],n1occ,0.0,Tp[0],n1occ);
-            C_DGEMM('T','N',nmo,n1occ,nso,-1.0,C1p[0],nmo,Tp[0],n1occ,1.0,Up[0],n1occ);
+            // TODO: We can simplify the code by converting these to transform calls, but that means allocating
+            //       temporary intermediates in a loop. In the (likely) event the allocation time cost is
+            //       negligible in comparison to the J/K build, simplify away.
+            C_DGEMM('N','N',nso,naocc,nso,1.0,J[2*a]->pointer()[0],nso,Caop[0],naocc,0.0,Tap[0],naocc);
+            C_DGEMM('T','N',nmo,naocc,nso,1.0,Cap[0],nmo,Tap[0],naocc,0.0,Uap[0],naocc);
```

Comment:
Change of constants from -1,1 to 1,0 deliberate?

Reply:
Yes.

In the old code, we backtransform four times and then add the pieces together. In the new code, we add the four pieces together and backtransform once. The job that those constants used to do is now fulfilled by `J[2*a]->add(J[2*a+1]); J[2*a]->scale(-1);` a few lines above.